core/src/main/java/org/apache/accumulo/core/data/KeyBuilder.java

```diff
@@ -0,0 +1,278 @@
+    protected boolean deleted = false;
+
+    final public ColumnFamilyStep<T> row(final T row) {
+      this.row = row;
+      return this;
+    }
+
+    @Override
+    final public ColumnQualifierStep<T> columnFamily(final T columnFamily) {
+      this.columnFamily = columnFamily;
+      return this;
+    }
+
+    @Override
+    final public ColumnVisibilityStep<T> columnQualifier(final T columnQualifier) {
+      this.columnQualifier = columnQualifier;
+      return this;
+    }
+
+    @Override
+    final public Build<T> columnVisibility(final T columnVisibility) {
+      this.columnVisibility = columnVisibility;
+      return this;
+    }
+
+    @Override
+    final public Build<T> timestamp(long timestamp) {
+      this.timestamp = timestamp;
+      return this;
+    }
+
+    @Override
+    public Build<T> deleted(boolean deleted) {
+      this.deleted = deleted;
+      return this;
+    }
+
+    @Override
+    public Key build() {
+      return this.build(true);
+    }
+
+    @Override
+    public Build<T> column(final T columnFamily, final T columnQualifier, final T columnVisibility) {
+      return this.columnFamily(columnFamily).columnQualifier(columnQualifier).columnVisibility(columnVisibility);
+    }
+
+    @Override
+    public ColumnVisibilityStep<T> column(final T columnFamily, final T columnQualifier) {
+      return this.columnFamily(columnFamily).columnQualifier(columnQualifier);
+    }
+  }
+
+  private static class TextKeyBuilder extends AbstractKeyBuilder<Text> {
+
+    private final Text EMPTY_TEXT = new Text();
+
+    @Override
+    public Key build(boolean copyBytes) {
+      Text columnFamily = (this.columnFamily == null) ? EMPTY_TEXT : this.columnFamily;
+      Text columnQualifier = (this.columnQualifier == null) ? EMPTY_TEXT : this.columnQualifier;
+      Text columnVisibility = (this.columnVisibility == null) ? EMPTY_TEXT : this.columnVisibility;
+      return new Key(row.getBytes(), 0, row.getLength(), columnFamily.getBytes(), 0, columnFamily.getLength(),
+          columnQualifier.getBytes(), 0, columnQualifier.getLength(), columnVisibility.getBytes(), 0,
+          columnVisibility.getLength(), timestamp, deleted, copyBytes);
+    }
+  }
+
+  private static class ByteArrayKeyBuilder extends AbstractKeyBuilder<byte[]> {
+
+    @Override
+    public Key build(boolean copyBytes) {
+      byte[] columnFamily = (this.columnFamily == null) ? EMPTY_BYTES : this.columnFamily;
+      byte[] columnQualifier = (this.columnQualifier == null) ? EMPTY_BYTES : this.columnQualifier;
+      byte[] columnVisibility = (this.columnVisibility == null) ? EMPTY_BYTES : this.columnVisibility;
+      return new Key(row, columnFamily, columnQualifier, columnVisibility, timestamp, deleted, copyBytes);
+    }
+  }
+
+  private static class CharSequenceKeyBuilder extends AbstractKeyBuilder<CharSequence> {
+
+    private final Text EMPTY_TEXT = new Text();
+
+    @Override
+    public Key build(boolean copyBytes) {
```

Comment:
No it doesn't for now because I'm still not sure if I can access the underlying array of a `CharSequence`. The issue is that not all types will allow to access their underlying bytes and with a common interface there is no way to set this flag only for the types that support this.
One solution would be to have another interface for types that support the use of the underlying bytes so that those types have the possibility to set the `copyBytes` flag and the others not but this complicates the things a little bit and it wouldn't play well with heterogeneous components. For instance, if I have

```java
byte r[];
String f;
Text q;

Key k = Key.newBuilder().row(r).family(f).qualifier(q).build();
```

Does build takes the flag `copyBytes` or not considering that `f` won't use the underlying array anyway? Shall we add the flag to each component method so that we can decide only for components that make sense? I think I rather prefer to have the flag regardless of the type with the idea what if the flag is set to `false` but the type doesn't support direct access to the underlying array, then the flag is ignored and bytes are copied.